src/nonunity/materials/CauchyStressFromNEML2.C

```diff
@@ -0,0 +1,200 @@
+
+InputParameters
+CauchyStressFromNEML2::validParams()
+{
+  InputParameters params =
+      NEML2SolidMechanicsInterface<ComputeLagrangianObjectiveStress>::validParams();
+  params.addClassDescription("perform the objective stress update using a NEML2 material model.");
+  params.addCoupledVar("temperature", "Coupled temperature");
+  return params;
+}
+
+CauchyStressFromNEML2::CauchyStressFromNEML2(const InputParameters & parameters)
+  : NEML2SolidMechanicsInterface<ComputeLagrangianObjectiveStress>(parameters),
+    // Inputs to the constitutive model
+    _mechanical_strain_old(nullptr),
+    _temperature(nullptr),
+    _temperature_old(nullptr)
+{
+  validateModel();
+
+  // Get the old mechanical strain if the NEML2 model need it
+  _mechanical_strain_old =
+      model().input().has_variable(strainOld())
+          ? &getMaterialPropertyOld<RankTwoTensor>(_base_name + "mechanical_strain")
+          : nullptr;
+
+  // Get the temperature if the NEML2 model need it
+  _temperature =
+      model().input().has_variable(temperature()) ? &coupledValue("temperature") : nullptr;
+
+  // Get the old temperature if the NEML2 model need it
+  _temperature =
+      model().input().has_variable(temperatureOld()) ? &coupledValueOld("temperature") : nullptr;
+
+  // Find the stateful state variables.
+  // A state variable is said to be stateful if it exists on the input's old_state axis as well as
+  // on the output's state axis.
+  if (model().input().has_subaxis("old_state"))
+    for (auto var : model().input().subaxis("old_state").variable_accessors(/*recursive=*/true))
+    {
+      _state_vars[var.on("state")] = &declareProperty<std::vector<Real>>(Moose::stringify(var));
+      _state_vars_old[var.on("old_state")] =
+          &getMaterialPropertyOld<std::vector<Real>>(Moose::stringify(var));
+    }
+}
+
+void
+CauchyStressFromNEML2::initialSetup()
+{
+  auto solver = _fe_problem.getNonlinearSystem().nonlinearSolver();
+  if (solver->residual_object || solver->jacobian || !solver->residual_and_jacobian_object)
+    mooseWarning("NEML2 material models are designed to be used together with "
+                 "residual_and_jacobian_together = true.");
+}
+
+void
+CauchyStressFromNEML2::initQpStatefulProperties()
+{
+  ComputeLagrangianObjectiveStress::initQpStatefulProperties();
+
+  // TODO: provide a mechanism to initialize the internal variables to non-zeros
+  for (auto && [var, prop] : _state_vars)
+    (*prop)[_qp] = std::vector<Real>(model().output().storage_size(var), 0);
+}
+
+void
+CauchyStressFromNEML2::computeProperties()
+{
+  try
+  {
+    // Allocate input
+    _in = neml2::LabeledVector::zeros(_qrule->n_points(), {&model().input()});
+
+    advanceStep();
+
+    updateForces();
+
+    if (model().implicit())
+      applyPredictor();
+
+    solve();
+  }
+  catch (...)
+  {
+    _console << "An error occurred during the evaluation of a NEML2 model."
+             << NEML2Utils::NEML2_help_message << std::endl;
+    throw;
```

Comment:
best make this a `mooseException(...)` call to play nice with the MooseLanguageServer diagnostics

Reply:
Sure. Just out of curiosity, how does the language server diagnostics work? Is there something I can read?

Reply:
> Is there something I can read?

Source code :rofl: 

The language server builds the app and catches exceptions to report those to the editor as issues with the current input. We have lose plans to improve that system to capture some metadata in the exception objects (like the name of the throwing object and it's position in the input file as separate data members so we don't have to parse the exception message to get those). Either way, a custom exception will not have that info in the message either, so string parsing would fail, too.